### PR TITLE
Fix build-info on MSVC.

### PR DESCRIPTION
**Pull request by dranger003**
Update `scripts/build-info.cmake` to fix erroring out using MSVC.
Below shows the error this update is fixing.

```
Z:\llama.cpp>md build && cd build && cmake -DLLAMA_CUBLAS=ON -DBUILD_SHARED_LIBS=ON ..
-- Building for: Visual Studio 17 2022
-- The C compiler identification is MSVC 19.37.32824.0
-- The CXX compiler identification is MSVC 19.37.32824.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/cmd/git.exe (found version "2.41.0.windows.3")
Microsoft (R) C/C++ Optimizing Compiler Version 19.37.32824 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

cl : Command line warning D9002 : ignoring unknown option '-dumpmachine'
cl : Command line error D8003 : missing source filename
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- Found CUDAToolkit: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.7/include (found version "11.7.64")
-- cuBLAS found
-- The CUDA compiler identification is NVIDIA 11.7.64
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.7/bin/nvcc.exe - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Using CUDA architectures: 52;61;70
-- CMAKE_SYSTEM_PROCESSOR: AMD64
-- CMAKE_GENERATOR_PLATFORM:
-- x86 detected
-- Configuring done (10.4s)
-- Generating done (0.3s)
-- Build files have been written to: Z:/llama.cpp/build
```

### CODE REVIEW

File: scripts/build-info.cmake [hunk 1]
```diff
@@ -28,43 +24,32 @@ if(Git_FOUND)
...
-    endif()
-endif()
-
-if(GIT_HEAD_RESULT EQUAL 0 AND GIT_COUNT_RESULT EQUAL 0)
     set(BUILD_COMMIT ${HEAD})
     set(BUILD_NUMBER ${COUNT})
 endif()
 
-execute_process(
-    COMMAND sh -c "$@ --version | head -1" _ ${CMAKE_C_COMPILER}
```
Comment (Cebtenzzre) [inline scripts/build-info.cmake:52]: Removing the use of `head` breaks BUILD_COMPILER for gcc:
```c
#define BUILD_COMPILER "gcc (GCC) 13.2.1 20230801
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE."
```

Comment (dranger003) [reply scripts/build-info.cmake:52]: Thanks for catching this I reverted this change.

File: scripts/build-info.cmake [hunk 2]
```diff
@@ -28,43 +24,32 @@ if(Git_FOUND)
...
-    OUTPUT_VARIABLE OUT
-    OUTPUT_STRIP_TRAILING_WHITESPACE
-    RESULT_VARIABLE RES
-)
-if (RES EQUAL 0)
+if(MSVC)
+    set(BUILD_COMPILER ${CMAKE_CXX_COMPILER_ID})
```
Comment (Cebtenzzre) [inline scripts/build-info.cmake:39]: Since we're checking the name of the C compiler below, we should use CMAKE_C_COMPILER_ID here for consistency.

Comment (dranger003) [reply scripts/build-info.cmake]: Thanks, updated for consistency.

**Comment by Cebtenzzre** [comment]
I tried with `mkdir build && cmake .. && cmake --build .` on Windows and did not get any useful compiler information. BUILD_COMPILER and BUILD_TARGET are both empty.

**Comment by dranger003** [comment]
> I tried with `mkdir build && cmake .. && cmake --build .` on Windows and did not get any useful compiler information. BUILD_COMPILER and BUILD_TARGET are both empty.

I think you meant `md` since this is on Windows. Below is the full output now, I updated `BUILD_TARGET` but `BUILD_COMPILER` is fine on my end - I am running this inside a VS dev command prompt.

```
Z:\>git clone -b build-info-msvc-fix https://github.com/dranger003/llama.cpp.git
Z:\>cd llama.cpp
Z:\llama.cpp>md build && cd build && cmake.exe ..
-- Building for: Visual Studio 17 2022
-- The C compiler identification is MSVC 19.37.32824.0
-- The CXX compiler identification is MSVC 19.37.32824.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/cmd/git.exe (found version "2.41.0.windows.3")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- CMAKE_SYSTEM_PROCESSOR: AMD64
-- CMAKE_GENERATOR_PLATFORM:
-- x86 detected
-- Configuring done (5.1s)
-- Generating done (0.3s)
-- Build files have been written to: Z:/llama.cpp/build

Z:\llama.cpp\build>type ..\build-info.h
#ifndef BUILD_INFO_H
#define BUILD_INFO_H

#define BUILD_NUMBER 1271
#define BUILD_COMMIT "9051f38"
#define BUILD_COMPILER "MSVC"
#define BUILD_TARGET "x64"

#endif // BUILD_INFO_H

Z:\llama.cpp\build>
```

**Comment by Cebtenzzre** [comment]
It shows up correctly at configure time, but when you run `cmake --build` it overwrites the file with one without compiler information.

**Comment by dranger003** [comment]
> It shows up correctly at configure time, but when you run `cmake --build` it overwrites the file with one without compiler information.

Interesting, I will need to do some research as I'm not overly familiar with cmake and I don't know why this is getting overwritten.

**Comment by dranger003** [comment]
@Cebtenzzre I have added a commit to pass-in the variables let me know what you think, thanks.

**Comment by Cebtenzzre** [comment]
Now I understand how @slaren ran into this:
```
_: line 1: --version: command not found
```
We need to pass CMAKE_C_COMPILER to the script as well, for use on non-Windows platforms.

**Comment by Cebtenzzre** [comment]
Seems to work fine on Linux and Windows. One other thing - cmake is able to identify the MSVC version:
```
-- The C compiler identification is MSVC 19.36.32537.0
```
Do you think we could make that appear in the build info? On Unix-like platforms, we use `--version` to get both the compiler name and version.

**Review by cebtenzzre — APPROVED** [review]

**Pull request merged**